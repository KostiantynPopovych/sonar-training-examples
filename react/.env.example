SONAR_URL=http://127.0.0.1:8084
SONAR_LOGIN=#auth token from the SonarQube instance
