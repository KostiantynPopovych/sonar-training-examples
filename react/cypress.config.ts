import { defineConfig } from 'cypress'

const baseUrl = 'http://localhost:3000';
const collectCoverage = 'CYPRESS_COVERAGE' in process.env;

export default defineConfig({
	video: false,
	...(collectCoverage ? {
		reporter: 'cypress-sonarqube-reporter',
		reporterOptions: {
			overwrite: true,
			outputDir: 'coverage/cypress/reports',
			useAbsoluteSpecPath: true,
			preserveSpecsDir: false,
			mergeOutputDir: 'coverage/cypress',
			mergeFileName: 'sonarqube-report.xml'
		},
	} : {}),
	env: {
		codeCoverage: {
			exclude: ['cypress/**/*.*', 'node_modules/**/*.*']
		}
	},
	e2e: {
		baseUrl,
		setupNodeEvents(on, config) {
			if (collectCoverage) {
				require('@cypress/code-coverage/task')(on, config);

				on('after:run', (results) => {
					return require('cypress-sonarqube-reporter/mergeReports')(results);
				});
			}
			return config;
		}
	}
})
