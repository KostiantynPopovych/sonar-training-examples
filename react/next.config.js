/** @type {import('next').NextConfig} */

const instrumentTheCode = 'INSTRUMENT_CODE' in process.env;

const nextConfig = {
  reactStrictMode: true,
  ...(instrumentTheCode ? {
    experimental: {
      swcPlugins: [
        [
          "swc-plugin-coverage-instrument", {}
        ]
      ]
    }
  } : {})
}

module.exports = nextConfig
