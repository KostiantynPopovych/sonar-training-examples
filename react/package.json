{
  "name": "react-example",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "dev:instrument": "cross-env INSTRUMENT_CODE=true next dev",
    "build": "next build",
    "start": "next start",
    "tests:unit": "jest",
    "tests:unit:watch": "jest --watch",
    "pretests:unit:coverage": "rimraf .next coverage/jest",
    "tests:unit:coverage": "jest --coverage",
    "cypress": "cypress open",
    "cypress:headless": "cypress run",
    "tests:e2e": "start-server-and-test dev http://localhost:3000 cypress",
    "tests:e2e-headless": "start-server-and-test dev http://localhost:3000 cypress:headless",
    "pretests:e2e:headless:coverage": "rimraf .next coverage/cypress .nyc_output",
    "tests:e2e:headless:coverage": "cross-env CYPRESS_COVERAGE=true start-server-and-test dev:instrument http://localhost:3000 cypress:headless",
    "test:both:coverage": "npm run tests:unit:coverage && npm run tests:e2e:headless:coverage",
    "update-sonar-project": "./update-sonar-project.sh"
  },
  "dependencies": {
    "next": "^13.1.1",
    "react": "^18.1.0",
    "react-dom": "^18.1.0"
  },
  "devDependencies": {
    "@cypress/code-coverage": "^3.10.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.2.0",
    "@types/jest": "^29.2.5",
    "@types/react": "18.0.9",
    "cross-env": "^7.0.3",
    "cypress": "^12.3.0",
    "cypress-sonarqube-reporter": "^1.12.0",
    "jest": "28.1.0",
    "jest-environment-jsdom": "^29.3.1",
    "jest-sonar": "^0.2.15",
    "rimraf": "^3.0.2",
    "sonarqube-verify": "^1.0.2",
    "start-server-and-test": "^1.15.2",
    "swc-plugin-coverage-instrument": "^0.0.14",
    "typescript": "4.6.4"
  }
}
