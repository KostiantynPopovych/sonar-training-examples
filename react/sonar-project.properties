# Source File Inclusions/Exclusions
sonar.sources=pages,styles
sonar.exclusions=/**/*.test.ts,/**/*.test.tsx

# Test File Inclusions/Exclusions
sonar.tests=__tests__,cypress/e2e
sonar.tests.inclusions=__tests__/**/*.test.ts,__tests__/**/*.test.tsx,cypress/e2e/*.cy.ts

# TypeScript Code coverage
sonar.javascript.lcov.reportPaths=coverage/jest/lcov.info,coverage/cypress/lcov.info
sonar.testExecutionReportPaths=coverage/jest/sonarqube-report.xml,coverage/cypress/sonarqube-report.xml
