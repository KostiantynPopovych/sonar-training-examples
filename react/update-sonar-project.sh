export $(cat .env | grep "^SONAR" | xargs)
sonarqube-verify
