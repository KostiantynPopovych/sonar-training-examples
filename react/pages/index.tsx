import Link from 'next/link'
import styles from '@/styles/shared.module.css'
import {FormEventHandler, useRef, useState} from "react";

export default function Home() {
  const inputRef = useRef<HTMLInputElement>(null);
  const [testedFrom, setTestedFrom] = useState('');

  const handleSubmit: FormEventHandler<HTMLFormElement> = (ev) => {
    ev.preventDefault();
    if (inputRef.current!.value === 'from-jest') {
      setTestedFrom('Tested by jest');
    } else if (inputRef.current!.value === 'from-cypress') {
      setTestedFrom('Tested by cypress');
    } else {
      setTestedFrom('Not tested');
    }
  }

  return (
    <div className={styles.container}>
      <main className={styles.main}>
        <form data-testid="testing-form" onSubmit={handleSubmit}>
          <input ref={inputRef} data-testid="testing-input" placeholder="test me" />
          <button data-testid="testing-button">Submit</button>
          <p data-testid="testing-p">{testedFrom}</p>
        </form>
        <h1>Home Page</h1>
        <p className={styles.description}>
          <Link href="/about">Go To The About Page &rarr;</Link>
        </p>
      </main>
    </div>
  )
}
