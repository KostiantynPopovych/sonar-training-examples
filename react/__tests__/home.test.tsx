import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import Home from '@/pages/index'

describe('Home', () => {
  it('renders link to about page', () => {
    render(<Home />)

    const homeTitle = screen.getByRole('heading', {
      name: /home page/i
    })

    const aboutLink = screen.getByRole('link', {
      name: /go to the about page/i,
    })

    expect(homeTitle).toBeInTheDocument()
    expect(aboutLink).toBeInTheDocument()
  })

  it('page should display tested by jest', async () => {
    render(<Home />)

    const input = screen.getByTestId('testing-input')
    const button = screen.getByTestId('testing-button')
    const p = screen.getByTestId('testing-p')

    fireEvent.change(input,  { target: { value: 'from-jest' } })
    fireEvent.click(button)

    await waitFor(() => {
      expect(p).toHaveTextContent(/tested by jest/i);
    });
  })

  /*
  * Uncomment this test to see that the coverage report will show
  * no untested paths, the test output is merged correctly and the coverage is 100%.
  */
  // it('should test untested path', async () => {
  //   render(<Home />)
  //
  //   const input = screen.getByTestId('testing-input')
  //   const button = screen.getByTestId('testing-button')
  //   const p = screen.getByTestId('testing-p')
  //
  //   fireEvent.change(input,  { target: { value: 'something' } })
  //   fireEvent.click(button)
  //
  //   await waitFor(() => {
  //     expect(p).toHaveTextContent(/not tested/i);
  //   });
  // })
})

it('renders homepage unchanged', () => {
  const { container } = render(<Home />)
  expect(container).toMatchSnapshot()
})
