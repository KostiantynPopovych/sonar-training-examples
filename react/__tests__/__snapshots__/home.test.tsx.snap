// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div
    class="container"
  >
    <main
      class="main"
    >
      <form
        data-testid="testing-form"
      >
        <input
          data-testid="testing-input"
          placeholder="test me"
        />
        <button
          data-testid="testing-button"
        >
          Submit
        </button>
        <p
          data-testid="testing-p"
        />
      </form>
      <h1>
        Home Page
      </h1>
      <p
        class="description"
      >
        <a
          href="/about"
        >
          Go To The About Page â†’
        </a>
      </p>
    </main>
  </div>
</div>
`;
